### PR TITLE
API examples overhaul & roundtrip tests

### PR DESCRIPTION
**Pull request by teh-cmc**
**Commit by commit**

This PR overhauls API examples to make them roundtrippable and checks those roundtrips on CI.

These roundtrips serve two purposes: A) the obvious one: they make sure that using our different SDKs in similar ways actually yield similar data, but more importantly B) they act as regression tests since it is highly unlikely that two or more tests written in different languages fail in the exact same way at the exact same time.

As I've painfully found out during the `MsgSender` migration, _a lot_ of our tests had bitrotten and simply did not work in one way or another anymore (in fact I ended up fixing yet another batch for this PR).
This PR should provide the foundations so that this doesn't happen again.

A nice side-effect of this is the introduction of the `_RERUN_TEST_FORCE_SAVE` environment variable, which forces all `RecordingStream`s instantiated across all 3 languages to write to an rrd file on disk rather than do whatever they were asked to do.
This makes testing things _much_ easier.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3204) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/3204)
- [Docs preview](https://rerun.io/preview/b04f8a8439f9a1ddfaa84c39637d17e75e79d338/docs) 
- [Examples preview](https://rerun.io/preview/b04f8a8439f9a1ddfaa84c39637d17e75e79d338/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: docs/code-examples/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# API Examples
+
+These examples showcase common usage of each individual Rerun `Archetype`s.
```
Comment (nikolausWest) [inline docs/code-examples/README.md:3]:
```suggestion
These examples showcase common usage of each individual Rerun `Archetype`.
```

File: docs/code-examples/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+Most of these example are automatically used as docstrings for the `Archetype` APIs in their respective SDKs.
+
+## Usage
+
+You can run each example individually using the following:
+
+- **Python**: `python <example_name>.py`, e.g. `python point3d_random.py`.
+- **Rust**: `cargo r -p code_examples --bin <example_name`, e.g.
```
Comment (jleibs) [inline docs/code-examples/README.md:12]:
```suggestion
- **Rust**: `cargo r -p code_examples --bin <example_name>`, e.g.
```

File: docs/code-examples/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  `cargo r -p code_examples --bin point3d_random`.
+- **C++**:
+  - `./docs/code-examples/build_all.sh` to compile all examples
+  - start a Rerun Viewer listening on the default port: `rerun`
+  - `./build/docs/code-examples/doc_example_<example_name>` to run, e.g.  `./build/docs/code-examples/doc_example_point3d_random`
```
Comment (jleibs) [inline docs/code-examples/README.md:17]: Given that these get generated in the `code-example` folder, do we need the `doc_example_` prefix?

File: docs/code-examples/annotation_context_connections.py [hunk 4]
```diff
@@ -1,28 +1,37 @@
 import rerun as rr
+import rerun.experimental as rr2
+from rerun.experimental import dt as rrd
```
Comment (jleibs) [inline docs/code-examples/annotation_context_connections.py:3]:
```suggestion
from rerun.experimental import dt as rrd
```

I think we should avoid `rrd` as the convention here due to name collision with `rrd` the file extension.

File: docs/code-examples/arrow3d_simple.rs [hunk 5]
```diff
@@ -28,3 +32,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
     rerun::native_viewer::show(storage.take())?;
     Ok(())
 }
+
+fn rnd(v: f64) -> f64 {
```
Comment (jleibs) [inline docs/code-examples/arrow3d_simple.rs:36]: The need for this detracts some from the example. If it's just here so we can get determinism on the round-trip-tests we should at least add a comment so we don't lead users down the path of thinking all angles and lengths need to be rounded for some reason.

File: docs/code-examples/depth_image_3d.py [hunk 6]
```diff
@@ -20,4 +20,4 @@
 )
 
 # Log the tensor, assigning names to each dimension
-rr.log_depth_image("world/camera/depth", image, meter=10000.0)
+rr.log_depth_image("world/camera/depth", image, tensor_id=np.repeat(0, 16).astype(np.uint8), meter=10000.0)
```
Comment (jleibs) [inline docs/code-examples/depth_image_3d.py:23]:
```suggestion
rr.log_depth_image("world/camera/depth", image, meter=10000.0)
```

I ran into this on the other image/tensor round-trip-tests as well.

This one is particularly bad. If a user copies this blindly it it's actually going to cause problems for them if they end up logging multiple tensors with the same id. 

I don't think we should leak this into our code-examples. We should either opt this one out of round-tripping or, better yet, we should implement a an exception for TensorId in the round-trip test framework.

File: docs/code-examples/depth_image_3d.rs [hunk 7]
```diff
@@ -13,27 +14,31 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
     let mut image = Array::<u16, _>::from_elem((200, 300).f(), 65535);
     image.slice_mut(s![50..150, 50..150]).fill(20000);
     image.slice_mut(s![130..180, 100..280]).fill(45000);
+
     let mut tensor = Tensor::try_from(image.as_standard_layout().view())?;
     tensor.meaning = TensorDataMeaning::Depth;
     tensor.meter = Some(10000.);
+    tensor.tensor_id = uuid::Uuid::nil().into();
```
Comment (jleibs) [inline docs/code-examples/depth_image_3d.rs:21]:
```suggestion
```
Again, this is going to cause problems if a user copies it.

File: docs/code-examples/depth_image_simple.py [hunk 8]
```diff
@@ -11,4 +11,4 @@
 rr.init("rerun_example_depth_image", spawn=True)
 
 # Log the tensor, assigning names to each dimension
-rr.log_depth_image("depth", image, meter=10000.0)
+rr.log_depth_image("depth", image, tensor_id=np.repeat(0, 16).astype(np.uint8), meter=10000.0)
```
Comment (jleibs) [inline docs/code-examples/depth_image_simple.py:14]: Same as above.

File: docs/code-examples/roundtrips.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+        if "rust" not in example_opt_out:
+            rust_output_path = run_roundtrip_rust(example, args.release, args.target, args.target_dir)
+
+            if "py" not in example_opt_out:
+                python_output_path = run_roundtrip_python(example)
+                run_comparison(python_output_path, rust_output_path, args.full_dump)
+
+            if "cpp" not in example_opt_out:
+                cpp_output_path = run_roundtrip_cpp(example, args.release)
+                run_comparison(rust_output_path, cpp_output_path, args.full_dump)
```
Comment (jleibs) [inline docs/code-examples/roundtrips.py:139]: Even if we opt out of *comparing* I suspect it's still helpful to do something like run the example and assert the RRD file is generated and not empty?

File: docs/code-examples/line_segments3d_simple.cpp [hunk 10]
```diff
@@ -5,18 +5,21 @@
...
-        {0.f, 0.f, 1.f},
-        {1.f, 0.f, 0.f},
-        {1.f, 0.f, 1.f},
-        {1.f, 1.f, 0.f},
-        {1.f, 1.f, 1.f},
-        {0.f, 1.f, 0.f},
-        {0.f, 1.f, 1.f},
-    };
-    rr_stream.log("segments", rr::LineStrips3D(points));
+    // TODO(#3202): I want to do this!
```
Comment (jleibs) [inline docs/code-examples/line_segments3d_simple.cpp:11]: Nit: we don't want to leak these kinds of TODOs into the embedded snippets in the doc-pages.

**Review by jleibs — APPROVED** [review]
Aside from the tensorid/meshid thing this looks great.

I also added an issue to check for TODOs in code examples (https://github.com/rerun-io/rerun/issues/3224). I think it's fine to have TODOs in main, we just need to be careful about not deploying them.

File: docs/code-examples/roundtrips.py (hunk 9 above)
Comment (teh-cmc) [reply docs/code-examples/roundtrips.py:139]: Right! Had the same thought earlier, let's go for it.

**Comment by teh-cmc** [comment]
- Will rebase this on top of #3230 to resolve all ID-related issues at once
- Will remove rounding in examples in favor of opting out

File: docs/code-examples/README.md [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
```
Comment (emilk) [inline docs/code-examples/README.md:18]: Please add a section about how to add new examples:

* Rust examples need to be manually added to `docs/code-examples/Cargo.toml`
* Maybe C++ examples need to be manually added somewhere?
* That we should add examples for all languages, and that they should match in what they do
* Their app ids must start with `rerun_example_`
* They should be short and simple
* …

File: docs/code-examples/line_segments3d_simple.cpp (hunk 10 above)
Comment (teh-cmc) [reply docs/code-examples/line_segments3d_simple.cpp:11]: I won't do anything about that for now; we'll cleanup during the pre-release doc pass.

#3224 is there to make sure we don't forget!

File: docs/code-examples/annotation_context_connections.py (hunk 4 above)
Comment (teh-cmc) [reply docs/code-examples/annotation_context_connections.py:3]: https://github.com/rerun-io/rerun/issues/3236

**Pull request merged**